sched/mutex/mutex_lock.c

```diff
@@ -0,0 +1,77 @@
+/****************************************************************************
+ * sched/mutex/mutex_lock.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/sched.h>
+#include <nuttx/mutex.h>
+
+#include <assert.h>
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nxmutex_lock
+ *
+ * Description:
+ *   This function attempts to lock the mutex referenced by 'mutex'.  The
+ *   mutex is implemented with a semaphore, so if the semaphore value is
+ *   (<=) zero, then the calling task will not return until it successfully
+ *   acquires the lock.
+ *
+ * Parameters:
+ *   mutex - mutex descriptor.
+ *
+ * Return Value:
+ *   This is an internal OS interface and should not be used by applications.
+ *   It follows the NuttX internal error return policy:  Zero (OK) is
+ *   returned on success.  A negated errno value is returned on failure.
+ *   Possible returned errors:
+ *
+ ****************************************************************************/
+
+int nxmutex_lock(FAR mutex_t *mutex)
+{
+  int ret;
+
+  DEBUGASSERT(!nxmutex_is_hold(mutex));
+  for (; ; )
+    {
+      /* Take the semaphore (perhaps waiting) */
+
+      ret = nxsem_wait(&mutex->sem);
+      if (ret >= 0)
+        {
+          mutex->holder = nxsched_gettid();
+          break;
+        }
+
+      if (ret != -EINTR && ret != -ECANCELED)
```

Comment:
I'm not sure how the caller should deal with `-ECANCELED`. Un libc implementation the mutex was remapped either to `nxsem_` ro to `sem_` API, so in user space the `sem_` API was used so `enter_cancellation_point()` and `leave_cancellation_point()` was applied. Now the path of `ECANCELED` is not clear to me

Reply:
Mutex is not a cancellation point, so don't need to check -ECANCELED, i'll delete it.